design/concise-tidy-design.Rmd

```diff
@@ -0,0 +1,215 @@
+---
+title: "Using tidy methods"
+author: "John Kirkpatrick"
+date: "07Jun2023"
+output: html_document
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+
+library(tidyverse)
+devtools::load_all()
+```
+
+## Introduction
+Here is a first draft implementation of broom-like methods for `crmPack, as suggested in [issue 323](https://github.com/Roche/crmPack/issues/323).
+
+### Principles
+The general principles I propose for tidying `crmPack` objects are as follows:
+
+* All slots that are not functions are converted to tibbles or a list of tibbles.
+* If the slot is scalar, the slot is converted to a 1x1 tibble.  This will ease downstream operations such as `row-bind`ing.
+* If the object being tidied contains multiple slots of (potentially) different lengths, the results is a list of tibbles.  The list may be nested to multiple levels.  (See, for example, `LogisticLogNormal`.)
+* The column names of the tidied tibble correspond to the slot names of the parent object.
+* If an object contains a vector or matrix of values that correspond to an "obvious" set of parameters, columns take on the names of those parameters and the tibble is augmented by a column named `Parameter`.  This most often happens in the mean vector and covariance matrix of `ModelParamsNormal`.
+* When the value of a slot has not been set, a zero-row tibble is returned.
+* When the value of a slot has scalar attributes, these attributes are added as a column of the tibble, whose name is the name of the attribute and whose value is the value of the attribute for every row of the tibble.  Vector attributes ill have to be handled on a case-by-case basis.
+* `tbl_<className>` is prepended to the class of the tidy tibble.
+
+### Exceptions
+*  Where a vector slot (or series of vector slots) define a range, then the naming convention described above is not followed.  Instead, columns named `min` and `max` define the extent of the range and a column named `Range` labels the range.
+
+### Please also read
+Please also read [Issue 407](https://github.com/Roche/crmPack/issues/407) regarding inconsistent slot names.
+
+## Implementation
+`tidy()` is a generic defined in `broom`.  An end-user of `crmPack` may not have installed `broom`.  Therefore, we need to define an appropriate geneic when necessary.
```

Comment:
I think the cleanest would be to import the `tidy` generic from the `broom` package.

Reply:
I agree, if you're happy to have `crmPack` Depend on `broom`.  I was trying to avoid that.

Reply:
no worries, since it is a standard package ok to depend on it